lib/quicrq.c

```diff
@@ -1828,22 +1913,41 @@ int quicrq_callback(picoquic_cnx_t* cnx,
         case picoquic_callback_stream_data:
         case picoquic_callback_stream_fin:
             /* Data arrival on stream #x, maybe with fin mark */
-            if (stream_ctx == NULL) {
-                /* Retrieve, or create and initialize stream context */
+            if ((stream_id & 2) == 0 && stream_ctx == NULL) {
+                /* Retrieve, or create and initialize stream context for control channel */
                 stream_ctx = quicrq_find_or_create_stream(stream_id, cnx_ctx, 1);
-            }
+                if (stream_ctx == NULL) {
+                    /* Internal error */
+                    (void)picoquic_reset_stream(cnx, stream_id, QUICRQ_ERROR_INTERNAL);
+                    return(-1);
+                }
 
-            if (stream_ctx == NULL) {
-                /* Internal error */
-                (void)picoquic_reset_stream(cnx, stream_id, QUICRQ_ERROR_INTERNAL);
-                return(-1);
-            }
-            else {
-                ret = quicrq_receive_stream_data(stream_ctx, bytes, length, (fin_or_event == picoquic_callback_stream_fin));
+                /* we consider unidir streams as warp/rush streams. may be we need something more explicit */
+                ret = quicrq_receive_stream_data(stream_ctx, bytes, length,(fin_or_event == picoquic_callback_stream_fin));
+
+            } else {
+                /* uni-directional stream are used for warp/rush flows */
+                if (cnx_ctx->control_stream == NULL) {
+                    /* Internal error, control stream should have been set up */
+                    (void) picoquic_reset_stream(cnx, stream_id, QUICRQ_ERROR_INTERNAL);
+                    return (-1);
+                }
+
```

Comment:
Why do you have `cnx_ctx->control_stream`? There has to be one control stream per media stream, just like we have per datagram. That control stream will be a property of the uni_stream, not a property of the connection.

Reply:
Per disucssion:
1 Stream Context - per media id and direction. MediaId is  controlled by receiver for uniqueness. Can work as even media_id by client and server creating the odd ones

Clean up when connection closes:
Having a list of uni_streams in  connection_ctx to handle non associated to control_stream

Reply:
OK.